Lib/pathlib.py

```diff
@@ -281,9 +281,14 @@ def _parse_parts(cls, parts):
         if altsep:
             path = path.replace(altsep, sep)
         drv, root, rel = cls._flavour.splitroot(path)
-        if drv.startswith(sep):
-            # pathlib assumes that UNC paths always have a root.
-            root = sep
+        if not root and drv.startswith(sep) and not drv.endswith(sep):
+            drv_parts = drv.split(sep)
+            if len(drv_parts) == 4 and drv_parts[2] not in '?.':
+                # e.g. //server/share
+                root = sep
+            elif len(drv_parts) == 6:
+                # e.g. //?/unc/server/share
+                root = sep
```

Comment:
The behavior isn't right when joining paths.

```pycon
>>> pathlib.Path('//./BootPartition') / 'Windows'
WindowsPath('//./BootPartitionWindows')
```

In this case, `PurePath.joinpath()` returns `self._from_parsed_parts(drv1, root1, parts1 + parts2)`. If `drv1` starts with a backslash but doesn't end with a backslash, and `root1` is empty, then `root1` should be set to backslash.

Reply:
Ack, right you are! Good spot!

Reply:
The `joinpath()` case will be solved when #101560 lands (it will start using the `_from_parts()` constructor and therefore `os.path.join()`).

Fixing `_make_child_relpath()` case (used by `iterdir()` and friends) isn't as simple as setting `root` in the right circumstances, as doing so can produce a `Path` object that is otherwise impossible to create, and that will compare unequal with other paths with the same string representation. I'm not even sure if this code path is possible, but I'm going to make it call `_from_parts()` instead to re-parse the joined path.